sklearn/preprocessing/_discretization.pyx

```diff
@@ -0,0 +1,55 @@
+import cython
+
+import numpy as np
+cimport numpy as np
+ctypedef np.float64_t DOUBLE  # This is simply `double`
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def binsearch(contcolumn, zero_interval, search_points):
+    # TODO: Add sparse array support
+
+    cdef DOUBLE[::1] _search_points = search_points.astype(float)
+
+    _search_points = np.insert(_search_points, 0, -np.inf)
+    _search_points = np.append(_search_points, np.inf)
+
+    cdef:
+        np.ndarray[DOUBLE, ndim=1, mode="c"] values = contcolumn.astype(float)
+        np.ndarray[DOUBLE, ndim=1, mode="c"] output = np.zeros(contcolumn.shape[0])
+        DOUBLE lower = <DOUBLE> zero_interval[0]
+        DOUBLE upper = <DOUBLE> zero_interval[1]
+        unsigned int i = 0, keyindex
+
+        # Need to have as many keys as search points, minus 1 to leave
+        # out np.inf, but plus 1 because our original offset is 1
+        long[::1] keys = np.arange(1, _search_points.shape[0])
+        DOUBLE value
+
+    with nogil:
+        for value_index in range(values.shape[0]):
+            value = values[value_index]
+            if lower <= value < upper:
+                output[i] = 0
+            else:
```

Comment:
Instead of a special condition for zeros here, this could be baked into the `keys` array.

`keys` could be designed to go from `[-k, ..., 0, ... +z]` ensuring that zeros map to zeros.
(The only important reasons why zeros map to zeros is so that we can safely iterate only over nonzeros when given a sparse matrix as input.)